### ISSUE TITLE
Address createCameraConfig race condition

### ISSUE THREAD

**Issue by nathanielrindlaub** [tnc-ca-geo/animl-api#112]
Sometimes while bulk-uploading images, because many animl-api lambdas have been concurrently spun up and are are working in tandem, multiple lambdas can be checking for existing camera configs for different images from the same camera, not find find an existing one, and then all create new camera configs at the same time, which results in duplicate camera config records. 

For example, the following two logs were from two separate lambdas, both checking for a camera config with the same serial number .004 seconds apart, which resulted in both lambdas creating identical camera configs. I'm not entirely sure what to do about this but it definitely is an issue.

```
2023-08-04T21:26:23.182Z	2dc73595-1c7a-4980-b97f-db4fb0a378f2 INFO couldnt find cameraConfig, so creating one...
2023-08-04T21:26:23.186Z	a6ac984d-9d83-4024-9178-acbc6c233c14 INFO couldnt find cameraConfig, so creating one...

**Comment by nathanielrindlaub** [comment]
I think we may be able to solve for this by adding a [multikey index](https://www.mongodb.com/docs/manual/core/index-multikey/) to ensure all `cameraConfigs` are unique within a `Project`:

```javascript
ProjectSchema.index({ 'cameraConfigs._id' : 1}, { unique: true } );
```

So in effect the racing lambdas would both not find an existing record and both try to create a new cameraConfig at the same time, but (hopefully) MongoDB would reject one of them. This is worth testing as I'm not 100% confident it would work. We'd also want to update the error handling of `createCameraConfig()` to ignore duplicate (?) errors thrown by the DB in that instance.

I also looked into Mongo's `findOneAndUpdate()` with an `upsert: true` flag (see SO post [here](https://stackoverflow.com/questions/33305623/mongoose-create-document-if-not-exists-otherwise-update-return-document-in)), but we don't really want to update the document if it is found, we want to do nothing.

**Comment by nathanielrindlaub** [comment]
False alarm re. the multikey index. Apparently the uniqueness is not enforced against the subdocuments within a given document but rather across other documents. So different `Projects` would not be allowed to have `cameraConfig` subdocuments with the same Id, but duplicate `cameraConfig`s would be permitted within individual `Projects`. That's the opposite of what we want.

From the MongoDB [docs](https://www.mongodb.com/docs/manual/core/index-unique/#std-label-unique-separate-documents):
> Because the constraint applies to separate documents, for a unique [multikey](https://www.mongodb.com/docs/manual/core/index-multikey/) index, a document may have array elements that result in repeating index key values as long as the index key values for that document do not duplicate those of another document. In this case, the repeated index entry is inserted into the index only once.

**Comment by nathanielrindlaub** [comment]
`$addToSet` seemed promising (see [docs](https://www.mongodb.com/docs/manual/reference/operator/update/addToSet/)), but it identifies duplicates by comparing at all of the fields in the subdocument you're trying to add to the array. This won't work for us because we create a default `cameraConfig.deployment` sub-subdocument when we call `createCamera()`, so if there's already an existing `cameraConfig` with a matching `_id`, `$addToSet` won't work b/c it's `cameraConfig.deployment._id` will be different than the one we're trying to add.

> If the value is a document, MongoDB determines that the document is a duplicate if an existing document in the array matches the to-be-added document exactly; i.e. the existing document has the exact same fields and values and the fields are in the same order. As such, field order matters and you cannot specify that MongoDB compare only a subset of the fields in the document to determine whether the document is a duplicate of an existing array element.

**Comment by nathanielrindlaub** [comment]
On second thought I'm a little perplexed how we're winding up with duplicate `cameraConfigs` and am not entirely sure it's race-condition related. If you take a look at the `createCamera()` logic (copied below), we first find the `Project`, then check the array of `cameraConfigs` to see if one exists with the same `_id` as the new camera, and if not, push it to the array and finally _save the entire project_. So even if two lambdas attempted to do this at the same time, the last one to save the `Project` would just overwrite the entire array of nested `cameraConfig` documents and there should never be any duplicates in there, right?

```javascript
        const [project] = await ProjectModel.getProjects([projectId], context);

        // make sure project doesn't already have a config for this cam
        const currCamConfig = project.cameraConfigs.find((cc) => idMatch(cc._id, cameraId));
        if (!currCamConfig) {
          console.log('couldnt find cameraConfig, so creating one...');
          project.cameraConfigs.push({
            _id: cameraId,
            deployments: [{
              name: 'default',
              timezone: project.timezone,
              description: 'This is the default deployment. It is not editable',
              editable: false
            }]
          });
          await project.save();
        }
        return project;
```

**Comment by nathanielrindlaub** [comment]
I think I may have found a solution using [`findOneAndUpdate()`](https://masteringjs.io/tutorials/mongoose/findoneandupdate) along with an aggregation pipeline. In [theory](https://www.mongodb.com/docs/manual/core/write-operations-atomicity/), a "write operation is [atomic](https://www.mongodb.com/docs/manual/reference/glossary/#std-term-atomic-operation) on the level of a single document, even if the operation modifies multiple embedded documents within a single document.", so by combining the query and the update into a single DB call we let MongoDB handle racing concurrent operations:

```javascript
  const updatedProject = await Project.findOneAndUpdate(
    { _id: projectId },
    [
      {
        $addFields: { camIds : '$cameraConfigs._id' }  // project a field that's an array of all cameraConfig _ids
      },
      {
        $set: {
          cameraConfigs: {
            $cond: {
              if: { $in: [cameraId, '$camIds'] },  // if the new cameraId is in the camIds array
              then: '$cameraConfigs',  // do nothing
              else: { $concatArrays: ['$cameraConfigs', [newCamConfig]] } // else add the new cameraConfig
            }
          }
        }
      }
    ],
    { returnDocument: 'after' } // returns the project after the update is complete
  );
```

I think it's worth reviewing all instances of using `find()` followed by and update and `save()` and refactor them using this approach if there's a risk of a race condition.
